consensus-peer/consensus/server/server.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/op/go-logging"
+
+	"golang.org/x/net/context"
+
+	google_protobuf "google/protobuf"
+)
+
+var logger = logging.MustGetLogger("server")
+
+type Server struct {
+	fabric_consensus.Executor
+
+	consensus fabric_consensus.Consenter
+	queueSize int
+
+	lock    sync.Mutex
+	deliver map[chan *consensus.Block]struct{}
+
+	executed *consensus.Block
+}
+
+type clientServer Server
+
+func New(queueSize int, conn *connection.Manager) *Server {
+	c := &Server{
+		queueSize: queueSize,
+		executed:  &consensus.Block{},
+		deliver:   make(map[chan *consensus.Block]struct{}),
+	}
+	consensus.RegisterAtomicBroadcastServer(conn.Server, (*clientServer)(c))
+	return c
+}
+
+func (c *Server) RegisterConsenter(consensus fabric_consensus.Consenter) {
+	c.consensus = consensus
+	c.Executor = executor.NewImpl(c.consensus, c, c)
+	c.Executor.Start()
+}
+
+// gRPC atomic_broadcast interface
+func (c *clientServer) Broadcast(ctx context.Context, msg *consensus.Message) (*google_protobuf.Empty, error) {
+	// XXX check ctx tls credentials for permission to broadcast
+	// XXX log broadcast
+	now := time.Now()
+	c.consensus.RecvRequest(&pb.Transaction{
+		Payload: msg.Data,
+		Timestamp: &google_protobuf.Timestamp{
+			Seconds: now.Unix(),
+			Nanos:   int32(now.UnixNano() % 1000000000),
+		},
+	})
+	return &google_protobuf.Empty{}, nil
+}
+
+func (c *clientServer) BroadcastStream(srv consensus.AtomicBroadcast_BroadcastStreamServer) error {
+	for {
+		msg, err := srv.Recv()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		c.consensus.RecvRequest(&pb.Transaction{
+			Payload: msg.Data,
+			Timestamp: &google_protobuf.Timestamp{
+				Seconds: now.Unix(),
+				Nanos:   int32(now.UnixNano() % 1000000000),
+			},
+		})
+		err = srv.Send(&consensus.Reply{})
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (c *clientServer) Deliver(_ *google_protobuf.Empty, srv consensus.AtomicBroadcast_DeliverServer) error {
+	// XXX check src tls credentials for permission to subscribe
+	// XXX log subscription
+	ch := make(chan *consensus.Block, c.queueSize)
+	c.lock.Lock()
+	c.deliver[ch] = struct{}{}
+	c.lock.Unlock()
+	for msg := range ch {
+		srv.Send(msg)
+	}
+	return nil
+}
+
+//
+func (c *Server) BeginTxBatch(id interface{}) error {
+	c.executed = &consensus.Block{}
+	return nil
+}
+
+func (c *Server) ExecTxs(id interface{}, txs []*pb.Transaction) ([]byte, error) {
+	msgs := make([]*consensus.Message, len(txs))
+	for i, tx := range txs {
+		msgs[i] = &consensus.Message{Data: tx.Payload}
+	}
+	c.executed.Messages = append(c.executed.Messages, msgs...)
+	return nil, nil
+}
+
+func (c *Server) CommitTxBatch(id interface{}, metadata []byte) (*pb.Block, error) {
+	b := c.executed
+	c.executed = &consensus.Block{}
+
+	c.lock.Lock()
+	out := c.deliver
+	c.lock.Unlock()
+
+	for ch := range out {
+		select {
+		case ch <- b:
+			break
+		default:
+			// ch is full, disconnect
+			c.lock.Lock()
+			delete(c.deliver, ch)
+			c.lock.Unlock()
+			close(ch)
+		}
+	}
```

Comment:
So you're still of the opinion that we should have a "don't proceed until X commit rule", rather than simply storing the blocks persistently locally? The local storage seems easier to me, and in the "As a service" model, I would almost definitely think providing replay would be desirable.